### PR TITLE
build(deps): bump reqwest from `v0.11.18` to `v0.12.2`

### LINKED ISSUE #4248: [Dependency Update] update `reqwest` from version `0.11.18` to `0.12.2`
`reqwest` recently released a breaking change (`v0.12.0`) and is quite a big change that could possibly fix the intermittent `connection reset` issue.

This is quite a big change and their changelog can be found here: https://github.com/seanmonstar/reqwest/blob/master/CHANGELOG.md

`reqwest` also hid the internal (not to be used) libraries that also includes `serde_json` which is used by `analytics` crate.

### PR DESCRIPTION
**Pull request by pixincreate**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [x] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR updates `reqwest` dependency version from `0.11.18` to `0.12.2` to address frequent disconnections and connection reset errors.

The major here is that we use `actix_web` along with `reqwest`. With `reqwest` releasing `0.12.0`, `http` crate version was bumped from `0.x` to `1.x` while on `actix_web v4.5.1` it is still on version `v0.2.9`. This version mismatch is what is handled here in this PR.

Also, in [webhook.rs](https://github.com/juspay/hyperswitch/blob/dd4afb4fa3828e8553152f489102b7856096f250/crates/router/src/core/webhooks.rs#L1216-L1232), it is not possible to directly convert `headers` that we get from `reqwest` to match with `actix_web` unless `actix_web` updates `http` version to `1.x`. Also, since `reqwest` does not support `HttpRequest` explicitly, we cannot directly use that either which led to manual conversion. (Open for discussions on this)

reqwest changelog can be found [here](https://github.com/seanmonstar/reqwest/blob/master/CHANGELOG.md) and notable breaking changes can be found [here](https://github.com/seanmonstar/reqwest/blob/master/CHANGELOG.md#v0120)

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Dependency update has been done in order to address frequent `connection_reset` errors.

Closes #4248

#### Basic tests

- Create { Merchant + API Key + MCA }
- Update API Key
- Make Payment 
- Refund

image

#### Postman tests

Once they succeeded, ran the postman collection for confirmation

image

image

#### Webhooks

image

#### Clickhouse

Set up clickhouse by following the guide [here](https://github.com/juspay/hyperswitch/tree/main/crates/analytics/docs/clickhouse)
http:localhost:8090

image
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/analytics/src/clickhouse.rs [hunk 1]
```diff
@@ -85,12 +85,14 @@ impl ClickhouseClient {
...
-                .change_context(ClickhouseError::ResponseError)?
-                .data)
+            let ckh_output: CkhOutput<serde_json::Value> =
+                match serde_json::from_str(response.text().await.unwrap_or_default().as_str()) {
+                    Ok(output) => output,
+                    Err(_) => {
+                        return Err(ClickhouseError::ResponseError.into());
+                    }
+                };
+            Ok(ckh_output.data)
```
Comment (pixincreate) [inline crates/analytics/src/clickhouse.rs:95]: This change was necessary as `reqwest` made `serde_json` to not be accessible as it is an internal feature which is used here at `analytics` crate (`.json`)

File: crates/router/src/core/webhooks.rs [hunk 2]
```diff
@@ -1213,6 +1213,24 @@ pub async fn webhooks_wrapper<W: types::OutgoingWebhookType, Ctx: PaymentMethodR
...
+    for (name, value) in actix_headers.iter() {
+        if let Ok(name) = reqwest::header::HeaderName::from_str(name.as_str()) {
+            if let Ok(value) = reqwest::header::HeaderValue::from_bytes(value.as_bytes()) {
+                reqwest_headers.insert(name, value);
+            }
+        }
+    }
+    reqwest_headers
+}
+
```
Comment (pixincreate) [inline crates/router/src/core/webhooks.rs:1233]: we can have a discussion on this.

File: crates/analytics/src/clickhouse.rs (hunk 1 above)
Comment (SanchithHegde) [reply crates/analytics/src/clickhouse.rs:95]: https://docs.rs/reqwest/latest/reqwest/struct.Response.html#method.json

> Available on crate feature `json` only.

Moreover, we have been pulling the `json` feature in other crates which use `reqwest`.

File: crates/router/src/core/webhooks.rs [hunk 3]
```diff
@@ -1213,6 +1213,24 @@ pub async fn webhooks_wrapper<W: types::OutgoingWebhookType, Ctx: PaymentMethodR
...
+    actix_headers: &actix_web::http::header::HeaderMap,
+) -> reqwest::header::HeaderMap {
+    let mut reqwest_headers = reqwest::header::HeaderMap::new();
+    for (name, value) in actix_headers.iter() {
+        if let Ok(name) = reqwest::header::HeaderName::from_str(name.as_str()) {
+            if let Ok(value) = reqwest::header::HeaderValue::from_bytes(value.as_bytes()) {
+                reqwest_headers.insert(name, value);
+            }
+        }
+    }
```
Comment (dracarys18) [inline crates/router/src/core/webhooks.rs:1230]: Can we atleast log the errors during conversion

**Review by dracarys18 — CHANGES_REQUESTED** [review]

File: crates/router/src/core/webhooks.rs (hunk 2 above)
Comment (pixincreate) [reply crates/router/src/core/webhooks.rs]: https://github.com/actix/actix-web/pull/3208

File: crates/router/build.rs
Comment (lsampras) [inline crates/router/build.rs:1]: is this needed?

**Review by lsampras — APPROVED** [review]

File: crates/router/build.rs
Comment (SanchithHegde) [reply crates/router/build.rs:1]: I'm not comfortable with us constantly bumping the number without attempting to address it.

Either revert this change, or if it is absolutely required, update the comment to reflect the updated number.

File: crates/pm_auth/Cargo.toml [hunk 4]
```diff
@@ -21,6 +21,7 @@ bytes = "1.4.0"
 error-stack = "0.4.1"
 http = "0.2.9"
 mime = "0.3.17"
+reqwest = "0.12.2"
```
Comment (SanchithHegde) [inline crates/pm_auth/Cargo.toml:24]: Can we avoid pulling one more dependency here, maybe bump `http` to `1.x` instead?

File: crates/router/build.rs
Comment (pixincreate) [reply crates/router/build.rs:1]: I started to get `stack overflow` on every run (stripe CI) resulting server shutdown until I bumped the min stack to 8 MiB.

**Review by dracarys18 — DISMISSED** [review]

**Review by ArjunKarthik — APPROVED** [review]
connector changes LGTM

**Comment by pixincreate** [comment]
After today's discussions, it has been decided to better wait before getting this PR merged as `actix_web` has an open PR that updates `http` crate version to `1.x` from `0.2.9` (which might take some time).

Once that is done, it will be easier for us to update the dependencies all together (`actix_web`, `http` and, `reqwest`) to their latest versions.

With that said, I'll be closing this PR to avoid any conflicts or confusion. I'll be raising yet another PR on behalf of this (just to keep the history clean) addressing the above mentioned changes.

**Comment by pixincreate** [comment]
I do not think this will ever be looked upon since:
- we are moving away from using `reqwest` in many places
- refactor that is being done

**Pull request closed without merging**